### ISSUE TITLE
预测

### ISSUE THREAD

**Issue by jie520520** [jie520520/jie#4]
import sys
import pandas as pd
import numpy as np
from PyQt5.QtWidgets import (
    QApplication, QWidget, QPushButton, QVBoxLayout, QLabel,
    QFileDialog, QTextEdit, QMessageBox
)
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from xgboost import XGBClassifier
from sklearn.preprocessing import StandardScaler

# ====================
# 特征构建函数
# ====================
def build_features(df, window_size=5):
    features = []
    labels = []
    periods = []

    for i in range(window_size, len(df)):
        window = df.iloc[i - window_size:i]
        current = df.iloc[i]
        feature = []

        for _, row in window.iterrows():
            feature.extend(row[['号码1', '号码2', '号码3', '号码4', '号码5', '号码6']].values)

        features.append(feature)
        labels.append(current['特码'])
        periods.append(current['期号'])

    return pd.DataFrame(features), pd.Series(labels), periods

# ====================
# 主窗口类
# ====================
class LotteryPredictor(QWidget):
    def __init__(self):
        super().__init__()
        self.setWindowTitle("六合彩预测程序（程序二）")
        self.setGeometry(100, 100, 600, 500)
        self.layout = QVBoxLayout()
        self.setLayout(self.layout)

        self.label = QLabel("请导入Excel文件")
        self.layout.addWidget(self.label)

        self.upload_button = QPushButton("上传数据文件")
        self.upload_button.clicked.connect(self.load_file)
        self.layout.addWidget(self.upload_button)

        self.predict_button = QPushButton("开始预测")
        self.predict_button.clicked.connect(self.predict)
        self.layout.addWidget(self.predict_button)

        self.result_text = QTextEdit()
        self.result_text.setReadOnly(True)
        self.layout.addWidget(self.result_text)

        self.df = None  # Excel 数据

    # ====================
    # 上传文件
    # ====================
    def load_file(self):
        file_path, _ = QFileDialog.getOpenFileName(self, "选择Excel文件", "", "Excel Files (*.xls *.xlsx *.xlsm)")
        if file_path:
            try:
                self.df = pd.read_excel(file_path, sheet_name="六合彩数据")
                self.label.setText(f"已加载数据，共 {len(self.df)} 行")
            except Exception as e:
                QMessageBox.warning(self, "错误", f"读取Excel出错：{str(e)}")

    # ====================
    # 预测逻辑
    # ====================
    def predict(self):
        if self.df is None:
            QMessageBox.warning(self, "错误", "请先上传数据文件！")
            return

        try:
            df = self.df.copy()
            df = df[['期号', '号码1', '号码2', '号码3', '号码4', '号码5', '号码6', '特码']]

            # 特征构造
            X, y, periods = build_features(df)

            # 🎯 修复 y 从 1~49 到 0~48
            y = y - 1

            # 特征归一化
            scaler = StandardScaler()
            X_scaled = scaler.fit_transform(X)

            # 模型训练
            rf = RandomForestClassifier(n_estimators=100, random_state=42)
            lr = LogisticRegression(max_iter=1000, random_state=42)
            xgb = XGBClassifier(n_estimators=100, use_label_encoder=False, eval_metric='mlogloss', random_state=42)

            rf.fit(X_scaled, y)
            lr.fit(X_scaled, y)
            xgb.fit(X_scaled, y)

            # 预测下期数据
            last_window = df.iloc[-5:][['号码1', '号码2', '号码3', '号码4', '号码5', '号码6']].values.flatten().reshape(1, -1)
            last_scaled = scaler.transform(last_window)

            rf_prob = rf.predict_proba(last_scaled)
            lr_prob = lr.predict_proba(last_scaled)
            xgb_prob = xgb.predict_proba(last_scaled)

            # 概率平均
            avg_prob = (rf_prob + lr_prob + xgb_prob) / 3

            # 取 Top-38 概率最高的号码
            top_k = 38
            top38 = np.argsort(avg_prob[0])[::-1][:top_k] + 1  # +1 还原号码
            top38.sort()

            next_period = df['期号'].max() + 1
            result = f"预测下一期：{next_period} 的 Top-{top_k} 号码为：\n" + ", ".join(map(str, top38))
            self.result_text.setText(result)

        except Exception as e:
            QMessageBox.critical(self, "预测失败", f"出错：{str(e)}")

# ====================
# 启动主程序
# ====================
if __name__ == "__main__":
    app = QApplication(sys.argv)
    window = LotteryPredictor()
    window.show()
    sys.exit(app.exec_())
